### ISSUE TITLE
can't close webpack-dev-server --progress with Ctrl + C

### ISSUE THREAD

**Issue by danburzo** [webpack/webpack-dev-server#1479]
* Operating System: macOS High Sierra (10.13.6)
* Node Version: v10.9.0
* NPM Version: 6.2.0
* webpack Version: 4.16.5
* webpack-dev-server Version: 3.1.5

- [x] This is a **bug**
- [ ] This is a **modification** request

### Code

Running the dev server with: `yarn webpack-dev-server --progress`. I'm not including webpack configs since seems to be reproducing on a variety of projects, so I assume it's something that happens regardless of the config? (Happy to provide more info otherwise).

### Expected Behavior

Ctrl + C stops the dev server.

### Actual Behavior

Ctrl + C does not do anything, dev server continues to compile modules.

_May be a regression of #860, #1360?_

**Comment by michael-ciniawsky** [comment]
Could you try with `v3.1.6` ? https://github.com/webpack/webpack-dev-server/releases/tag/v3.1.6 (#1432)

**Comment by danburzo** [comment]
Woo-hoo! Missed that there's 3.1.6 available, works like a charm with it. Thank you! :-)

**Comment by qkdreyer** [comment]
Still an issue with [`v3.1.9`](https://github.com/webpack/webpack-dev-server/releases/tag/v3.1.9)

**Comment by SebT** [comment]
Having the issue on Linux and Mac with webpack-dev-server `v3.1.8`.

**Comment by evilebottnawi** [comment]
Please create minimum reproducible test repo with steps

**Comment by superhawk610** [comment]
@SebT I've also had this issue, specifically when using the `-theme react` preset from the ReasonReact starter. I'm working on getting reproducible steps - it doesn't seem to happen in a dry environment, so I imagine it's some conflict with another file watcher that occurs after some period of use.

I'll report back if I get anything useful.

**Comment by superhawk610** [comment]
The error (in this case, at least) appears to occur when the dev server gets stuck compiling. `Ctrl+C` at this point doesn't actually end any processes - all 3 processes listed above remain active after sending `SIGINT` to the yarn process.

After further attempts, this still doesn't reproduce the error consistently :/ FWIW, I'm using VSCode with the Reason/OCaml extension.

**Comment by dcorb** [comment]
One developer in our team is having the same issue (the rest of the team not). So it's hard to offer a reproducible test.

**Comment by evilebottnawi** [comment]
If somebody can reproducible test please post how here, we can fix it asap

**Comment by Krzychek** [comment]
> The error (in this case, at least) appears to occur when the dev server gets stuck compiling. `Ctrl+C` at this point doesn't actually end any processes - all 3 processes listed above remain active after sending `SIGINT` to the yarn process.

The same happens for me on Linux (up-to-date ArchLinux). Also it drains 100% of one CPU core.

**Comment by rhys-vdw** [comment]
Can confirm we're still experiencing this at webpack-dev-server 3.1.10, macOS 10.13.6

**cc** @nfm confirm whether you're still experiencing this one?

**Comment by evilebottnawi** [comment]
Please create minimum reproducible test repo

**Comment by evilebottnawi** [comment]
@danburzo What os i need to reproduce?

**Comment by rhys-vdw** [comment]
OS:

```
-> sw_vers
ProductName:	Mac OS X
ProductVersion:	10.13.6
BuildVersion:	17G3025
```

Using the above repo:

![wds-kill-failure](https://user-images.githubusercontent.com/822105/49077459-70b5ae00-f28f-11e8-82e3-e61ddd375f12.gif)

Every time the progress bar is duplicated I am pressing `^C`. Also I need to press it to bring the prompt back at the end.

**Comment by rhys-vdw** [comment]
(Note that @nfm is experiencing this on Ubuntu so it's not just macOS)

**Comment by evilebottnawi** [comment]
@danburzo Awesome, reproduce problem on `linux` (`Ubuntu`) too

**Comment by evilebottnawi** [comment]
One interesting thing - it is happens when you try to `require('webpack')`, if you change this on `console.log` all works fine

**Comment by evilebottnawi** [comment]
Problem found, we use `server.close` before run `process.exit` (close all connections, watchers and etc stuff). `server.close` use `webpackDevServer.close` method, what use `watching.close` (when you in not lazy mode - using `webpack-dev-server`), but `watching.close` wait until all was compiled. 

What we need to fix all problem with `CRTL+C`:
1. Implement `Watching.kill()` method in webpack
2. Add `kill` option to `webpackDevServer.close` (by default `false`)
3. Use `webpackDevServer.close(cb, true)` in `Server.close`.

If somebody have time to help with this issue PRs welcome

**Comment by bhavya9107** [comment]
I want to work on this issue. How to get started?

**Comment by evilebottnawi** [comment]
@bhavya9107 

> 1. Implement Watching.kill() method in webpack

Here https://github.com/webpack/webpack/blob/master/lib/Watching.js

**Comment by bhavya9107** [comment]
What is expected behavior of kill method?

**Comment by evilebottnawi** [comment]
@bhavya9107 immediately kill watch on webpack side

**Comment by bhavya9107** [comment]
why can't we just process.exit()

**Comment by bhavya9107** [comment]
if we want to close a specific process then we need specific close and kill methods, otherwise process.exit() is fine according to me. What you say @evilebottnawi

**Comment by evilebottnawi** [comment]
@bhavya9107 because when you `process.exit()` you kill process immediately, but in this moment somebody/something can write file to fs or do other async call and it is bad

**Comment by caolvchong** [comment]
still in 3.2.1
my solution:
```js
// webpack.development.js
const kill = require('tree-kill')

const pid = process.pid
process.on('SIGINT', function () {
  kill(pid, 'SIGKILL', function () {})
})
```

**Comment by jayfunk** [comment]
So is this bug fixed and supposed to be in the latest release?

**Comment by evilebottnawi** [comment]
@caolvchong it can kill some processes what writing files or other stuff what can break you app

**Comment by Hauuguu** [comment]
Any update on this?

**Comment by Loonride** [comment]
@evilebottnawi A bug was introduced when the `setupExitSignals` helper was added, related to this issue. The problem is that `server` is getting passed by value as undefined into the `setupExitSignals` function, and the function will never see the defined server object:

https://github.com/webpack/webpack-dev-server/blob/a28800dbe3b490360fd41c41802483c14e7d01db/bin/webpack-dev-server.js#L21-L23

Should we fix this on `master` or `next`?

**Comment by evilebottnawi** [comment]
@Loonride let's fix it in master

**Comment by hiroppy** [comment]
Can we close this issue? /ccc @evilebottnawi @Loonride

**Comment by evilebottnawi** [comment]
@hiroppy need manual test https://github.com/webpack/webpack-dev-server/issues/1479#issuecomment-442008093

**Comment by hiroppy** [comment]
fm... It doesn't seem fixed using webpack-dev-server@3.8.0. I'll check this issue.

**Comment by Loonride** [comment]
@hiroppy My fix did not solve the issue, it only changed the behavior slightly to remove a related bug. This still has to be done: https://github.com/webpack/webpack-dev-server/issues/1479#issuecomment-442492922

**Comment by hiroppy** [comment]
oh, ok, thanks.

**Comment by javaxiu** [comment]
have you guys found this error when press ctrl+C
```bash
if (serverData.server) {
                     ^
TypeError: Cannot read property 'server' of undefined
    at process.on (/Users/liangjiawen/workspace/recore/test/node_modules/webpack-dev-server/lib/utils/setupExitSignals.js:8:22)
    at process.emit (events.js:194:15)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
```
webpack-dev-server@3.9.0

**Comment by evilebottnawi** [comment]
@javaxiu good catch, can you send a PR with fix - `if (serverData && serverData.server)`?

**Comment by ndbroadbent** [comment]
This is still happening for me in webpack-dev-server `3.9.0` (and webpack `4.41.2`.) When I press Ctrl+C, I get returned to the shell prompt, but I still see the WebpackBar progress until the compilation has finished. 

@caolvchong's workaround works:

```
const kill = require('tree-kill')

const pid = process.pid
process.on('SIGINT', function () {
  kill(pid, 'SIGKILL')
})
```

**Comment by evilebottnawi** [comment]
will be fixed for webpack@5

**Comment by cben** [comment]
Seeing the same happening with webpack-dev-server 3.10.1, webpack 4.41.5 on Linux, compilation progress continues.
FWIW, <kbd>Ctrl</kbd>+<kbd>\ </kbd> (sends SIGQUIT) works immediately for me.  (same caveats apply as other "kill" solutions)

**Comment by creativemacmac** [comment]
Hey there folks I also seem to have this issue when running gatsbyjs or nextjs projects. I am on a windows 10 machine using webstorm. ctrl c will work a few times but then gets stuck. is there any permanent solution to it?

**Comment by bencresty** [comment]
Still the same issue in Webpack 4.46.0 + webpack-dev-server 3.11.2 on Win 10. Pretty frustring to not being able to stop a process when `--progress` is true and have to wait for it everytime.
Can't switch to Webpack 5, unfortunately, but anybody perhaps knows if this is already fixed in Webpack 5?

**Comment by alexander-akait** [comment]
Other reproducible example

**Comment by PikachuEXE** [comment]
@alexkuz 
You mentioned that [the bug is on webpack side](https://github.com/webpack/webpack-dev-server/issues/3412#issuecomment-866325121).

Is there anything we can do beside waiting? (investigating, filing an issue, figuring out a workaround/fix...)

**Comment by alexander-akait** [comment]
It should be fixed for webpack v5 and webpack-dev-server v4

**Comment by alk-sdavid** [comment]
I migrated the big fat app of my company to webpack-dev-server@4.1.1, and now it's quite slow to close the dev server (~10 seconds):


https://user-images.githubusercontent.com/50570955/132487635-e42d5446-eb9a-4985-b92b-206c3dd49f42.mp4

I didn't had this slowness with webpack-dev-server@3

webpack@5.52.0 + webpack-dev-server@3.11.2 = fast closing
webpack@5.52.0 + webpack-dev-server@4.1.1 = slow closing

**Comment by alexander-akait** [comment]
@alk-sdavid Do you use `cache`?

**Comment by alk-sdavid** [comment]
@alexander-akait Yes, but it's the same behavior with `cache: true` or `cache: false`. Sorry I can't give you a reproducible example since it's a private and confidential repo, and I don't have this issue with little projects with few files, so maybe maybe it's because it's a fat project... 🤷 Notice that I reproduce this slowness with `watch: false`.

**Comment by alexander-akait** [comment]
Can you try change https://github.com/webpack/webpack-dev-server/blob/master/lib/Server.js#L1090
```js
signals.forEach((signal) => {
        process.on(signal, () => {
          this.stopCallback(() => {
            // eslint-disable-next-line no-process-exit
            process.exit();
          });
        });
      });
```

on

```js
signals.forEach((signal) => {
        process.on(signal, () => {
          this.stopCallback(() => {
            this.compiler.close(() => { 
              // eslint-disable-next-line no-process-exit
              process.exit();
            })
          });
        });
      });
```

problems somewhere here, you can add on each level `console.log("Debug")` and look where it freeze

**Comment by alk-sdavid** [comment]
Ok I did this:

```js
signals.forEach((signal) => {
  process.on(signal, () => {
    console.time('stop');
    console.log('[DEBUG] on', signal);
    this.stopCallback(() => {
      console.timeLog('stop', '[DEBUG] stopCallback');
      this.compiler.close(() => {
        console.timeLog('stop', '[DEBUG] process.exit()');
        // eslint-disable-next-line no-process-exit
        process.exit();
      })
    });
  });
});
```
And so it takes 10sec to exec the stop callback:
```
^C[DEBUG] on SIGINT
99% end closing watch compilation stop: 10.456s [DEBUG] stopCallback
stop: 10.457s [DEBUG] process.exit()
```

Notice that I see `99% end closing watch compilation` at the same time as `[DEBUG] stopCallback`

**Comment by alexander-akait** [comment]
Can you try to disable `profile` plugin?

**Comment by alk-sdavid** [comment]
`profile: false` does not change anything. I'm using webpack@5.52.0. I dug a bit in the close method, and I logged this:

```
[DEBUG] on SIGINT
stop: 0.48ms [DEBUG] stop()
stop: 4.438ms [DEBUG] close websocket server
stop: 5.256ms [DEBUG] websocket server closed
stop: 5.335ms [DEBUG] close server
stop: 5.732ms [DEBUG] server closed
stop: 5.804ms [DEBUG] close middleware
99% end closing watch compilation stop: 10.587s [DEBUG] middleware closed
stop: 10.587s [DEBUG] stopCallback
stop: 10.588s [DEBUG] process.exit()
```

So the slowness looks to come from the middleware (webpack-dev-middleware), so sorry about that, it looks like it's not coming from webpack-dev-server 😅

**Comment by alexander-akait** [comment]
Do you watch many files? But I have idea hot we will fix it:
1. `CTRL + C` first press - graceful shutdown
2. `CTRL + C` second press - process.exit

Anyway, closing `middleware` is slower, need investigate, maybe you can debug deeply?

**Comment by alk-sdavid** [comment]
Yes I'm watching a lot of files (even if I ignored node_modules and test files) but `watch: false` should fix this issue but it's still slow 🤔 anyway I'm debugging it more deeply, and it might be a webpack issue, I'll have a look on this and I'll open ticket if I find something relevant.

**Comment by alexander-akait** [comment]
@alk-sdavid yep, feel free to feedback, this problem is not hard to debug, any additional information will help use too to understand there is the problem

**Comment by alexander-akait** [comment]
hm, we have built-in webpack watching https://github.com/webpack/watchpack/blob/master/lib/watchpack.js#L252 (try to debug it too, thanks) and chokidar (used due glob usage), does `watcher.close` means webpack built-in watch?

**Comment by alk-sdavid** [comment]
My logs where here https://github.com/webpack/webpack/blob/main/lib/Watching.js#L445

**Comment by alexander-akait** [comment]
Can you try to debug `watchpack`? Something wrong with close polling resources I think

**Comment by alk-sdavid** [comment]
yes sorry I try to debug this more and more deeply between my meetings 😅

**Comment by alexander-akait** [comment]
Feel free to feedback :+1:

**Comment by alk-sdavid** [comment]
So I have 23551 fileWatchers to close here https://github.com/webpack/watchpack/blob/master/lib/watchpack.js#L255, I guess it's a lot 😅 but I don't really have so much file in my project, it should be something like ~5000 files. When I look at paths of these fileWatchers, I can see that 1 real file produces many files, for example:

from 1 real file named `.../admin/constants/index.js`, it produces 10 more file watchers for:

```
.../admin/constants.mjs
.../admin/constants.js
.../admin/constants.jsx
.../admin/constants.ts
.../admin/constants.tsx
.../admin/constants.wasm
.../admin/constants.json
.../admin/constants/package.json
.../admin/constants/index
.../admin/constants/index.mjs
```
I guess there is a technical reason to this which might talk to you, isn't it?

**Comment by alexander-akait** [comment]
> I guess there is a technical reason to this which might talk to you, isn't it?

Yes, we need watch these files, because you can add them late...

**Comment by alk-sdavid** [comment]
The slowness is coming from this line https://github.com/webpack/watchpack/blob/master/lib/watchEventSource.js#L75 where the NodeJS FSWatcher is closed. In my case, I count 1219 executions of FSWatcher.close, if I comment the line: it's fast 😿

**Comment by alexander-akait** [comment]
Can you debug this too https://github.com/webpack/watchpack/blob/master/lib/watchEventSource.js#L188?

**Comment by alk-sdavid** [comment]
yes I passed here https://github.com/webpack/watchpack/blob/master/lib/watchEventSource.js#L194 then https://github.com/webpack/watchpack/blob/master/lib/watchEventSource.js#L75

So I can now reproduce the slowness on my laptop:

```js
const fs = require('fs');
const watchers = [];

for (let n = 0; n < 1200; n += 1) {
  watchers.push(fs.watch(process.cwd()));
}

console.time('close');
for (let w of watchers) {
  w.close();
}
console.timeEnd('close');
```

These fs.FSWatcher instances are slow to close (~4.7s on my laptop)

**Comment by alexander-akait** [comment]
Big thanks, we will investigate this in near future

**Comment by laverdet** [comment]
I'm working around this now by adding this to my webpack.config.js:

```js
const { watch } = require('fs');

// Ctrl + C is super duper slow because it has to close a bunch of fs.watch handles. Let's just skip
// all that.
// https://github.com/webpack/webpack-dev-server/issues/1479
process.once('SIGINT', () => {
    Object.getPrototypeOf(watch('.')).close = () => {};
});
```

**Comment by alexander-akait** [comment]
We will fix it in near future, anyway not graceful closing can corrupt file caches

**Comment by raine** [comment]
Checked watchpack's `this.fileWatchers` upon closing and for me there are 75k files watched. Many of which are in `node_modules`.

Any idea what's the point of watching those files and what is adding them? Thanks.

**Comment by alexander-akait** [comment]
Due resolving logic, I am thinking long closing is a bug in Node.js, What is os?

**Comment by raine** [comment]
I'm on macOS Catalina.

I added `**/node_modules` to `watchOptions.ignored`. That removed around half of the watchers and closing is faster.

Now that I'm checking the watched files in `this.fileWatchers`, I'm seeing that many files are watched with multiple extensions, even if they don't exist.

```
  ...
  "app/scripts/client/index.tsx", // <-- Only this actually exists in file system
  "app/scripts/client/index",
  "app/scripts/client/index.web.js",
  "app/scripts/client/index.js",
  "app/scripts/client/index.json",
  "app/scripts/client/index.ts",
```

Is this normal behavior? Seems to me like there are useless watchers but I don't understand this library.

**Comment by alexander-akait** [comment]
Yes, because you can add `index.ts` and we need to watch these adding, for example you have `index.ts`, then you want to use `index.tsx`, if you add them we need rerun compilation, same for other

**Comment by niksy** [comment]
Seeing same issue in Webpack 5.53.0, macOS Big Sur, M1, Node 12.22.5.

**Comment by alexander-akait** [comment]
I have idea how to improve this, there are two approach:
- one CTRL+C - gracefully shutting down
- second CTRL + C - force process closing

So you can chose wait or close immediately

**Comment by raine** [comment]
How realistic case is it that force exit corrupts webpack cache or something like that?

**Comment by alexander-akait** [comment]
In stage when webpack try to write caches, it is based on many metrics, hard to say, but I think in this case webpack should skip broken cache and try to create new

**Comment by laverdet** [comment]
> How realistic case is it that force exit corrupts webpack cache or something like that?

I'd eat my hat if leaving a dangling watch handle corrupts any file cache. My hack above just leaves the watch handles open, it doesn't touch anything else. The operating system is more than capable of cleaning our mess up.

**Comment by alexander-akait** [comment]
@laverdet yep hack above will be corrupt cache, it just fix Node.js problems with slow closing files.

Here two problems:
1. Bug on our side, we don't close compiler - https://github.com/webpack/webpack-dev-server/pull/3880, also we should ask developer to close gracefully or force
2. Bug on Node.js side with slow file closing

**Comment by laverdet** [comment]
The offer on eating my hat still stands. I just can't imagine a scenario in which a file would be corrupted by a dangling inotify handle. In fact, I'd go so far to say that any such corruption would warrant a CVE in the Linux kernel followed by a hushed patch by Google's top minds.

On the other hand calling process.exit will almost *definitely* cause partially written files to be flushed to the filesystem since the process terminates with open descriptors (of the mode == "w" kind).

I have a great deal of respect for your contributions to Webpack but I do think that leaving the watch handles open is a safer workaround in this case.

**Comment by alexander-akait** [comment]
There is another problems - developers middlewares, we can't know it is safe to close immediately or we need to do graceful shutdown, same for custom client/web socket server implementations, I think we need report problem with slow closing to Node.js. Anyway in the next release double pressing CTRL + C will close immediately process, so developer can choose what is better

**Comment by alexnix** [comment]
Hi, I am having an issue that is probably related to this.
In my setup I am calling `new WebpackDevServer(...).start()` from a script. Issue is Cmd+C (SIGINT) stops it immediately (at least in the shell it looks like so, it shows a prompt again, as if the previous command stopped), but then addition output from webpack dev server appears (the line about gracefully shutting down). At this point the terminal is in a state where non-asci characters -up/down arrows for example- appear as ^[OA^[OB). I can press Cmd+C again and it all works well, but I am just wondering why this happens (I mean the terminal wired state, not the double Cmd+C which I understand, by reading this thread).

I suspect that my script that calls `new WebpackDevServer(...).start()` should somehow also wait for it to stop on SIGINT...? so i tried:
```
process.once('SIGINT', async () => {
    devServer.stop().then(() => {
        process.exit(0);
    });
});
```

but it did not work.

Strangely, if I send SIGINT by `kill -SIGINT <pid>`, then it works well: the gracefully shutting messages appears and the shell does not change back to an input before those watches are closes.

Any suggestions?

**Comment by alexnix** [comment]
Hi, I am having an issue with might be related.

I start WebpackDevServer from a script, but when I Cmd+C to close it, the script stops immediately (at least the terminal looks like it did), but then not only additional output appears from webpack (the line with "gracefully shutdown") but also, if i try to input characters such as up/down arrows they appear as `[OB^[OB`.

I am wondering if i need to wait for webpack in anyway from within my script. 

I tried but did not work.
```
process.on('SIGINT', () =>{
    devServer.stop().then(() => {
        process.exit(0);
    });
});
```

Maybe it is work mentining that if i send the SIGINT from `htop` or `kill` it does works as expected: it prints the "gracefully shutdown message" then waits for the file watches to close, and even prints a "done" afterwards.

In your opinion, is this related to the webpack - script interaction, or is it a terminal/zsh related issue (give that SIGINT is handled corect when coming from a source different than Cmd+C) ?

**Comment by laverdet** [comment]
@alexnix what you are experiencing is a common issue when terminating processes which use vterm escape sequences, for example colors in Webpack's output. You can fix your terminal by running [`reset`](https://linux.die.net/man/1/reset) or by just opening a new terminal.

**Comment by alexnix** [comment]
Thanks for the clarification. 
So you are basically saying this issue is not related to my code, but rather some common issue with terminals ?

**Comment by laverdet** [comment]
Right, it's totally normal. Just remember the `reset` command for situations like this.

**Comment by DominicTobias** [comment]
The only way it exits for me is if I run it webpack directly from CLI, if I put it in a shell script it prints "Gracefully shutting down" but then never does, regardless of how many watchers there are. Also tried things like `pkill -P $$` on EXIT but no avail (using 4.15.1 and webpack5)

**Comment by jtbandes** [comment]
Adding `watchOptions: { ignored: "**/node_modules/**" }` made shutdown much faster for me (from several seconds to nearly instant).

I believe I figured out the reason that "first ^C graceful shutdown, second ^C force close" was not actually working — it's because the server was running `fileWatcher.close()` [in a tight loop](https://github.com/webpack/watchpack/blob/1a1feae5db7d7955cfdb32d7457a0454820fa38c/lib/watchpack.js#L310), which prevented the SIGINT handler from executing a second time. If these were closed asynchronously, the handler would be able to run again and call process.exit().
